dali/kernels/signal/fft/fft_cpu_impl_ffts.cc

```diff
@@ -0,0 +1,171 @@
+  constexpr bool is_real_out = std::is_same<OutputType, float>::value;
+  DALI_ENFORCE((is_complex_out && args.spectrum_type == FFT_SPECTRUM_COMPLEX)
+            || (is_real_out && args.spectrum_type != FFT_SPECTRUM_COMPLEX),
+    "Output type should be complex<float> or float depending on the requested spectrum type");
+
+  transform_axis_ = args.transform_axis >= 0 ? args.transform_axis : Dims-1;
+  DALI_ENFORCE(transform_axis_ >= 0 && transform_axis_ < Dims,
+    make_string("Transform axis ", transform_axis_, " is out of bounds [0, ", Dims, ")"));
+
+  const auto n = in.shape[transform_axis_];
+  auto nfft = args.nfft > 0 ? args.nfft : n;
+
+  KernelRequirements req;
+  auto out_shape = in.shape;
+
+  ScratchpadEstimator se;
+  se.add<float>(AllocType::Host, size_in_buf(nfft),  32);
+  se.add<float>(AllocType::Host, size_out_buf(nfft), 32);
+  req.scratch_sizes = se.sizes;
+
+  if (args.spectrum_type == FFT_SPECTRUM_COMPLEX) {
+    out_shape[transform_axis_] = nfft;
+  } else {
+    out_shape[transform_axis_] = nfft / 2 + 1;
+  }
+  req.output_shapes = {TensorListShape<DynamicDimensions>({out_shape})};
+
+  if (!plan_ || nfft != nfft_) {
+    if (can_use_real_impl(nfft)) {
+      plan_ = {ffts_init_1d_real(nfft, FFTS_FORWARD), ffts_free};
+    } else {
+      plan_ = {ffts_init_1d(nfft, FFTS_FORWARD), ffts_free};
+    }
+    DALI_ENFORCE(plan_ != nullptr, "Could not initialize ffts plan");
+    nfft_ = nfft;
+  }
+
+  return req;
+}
+
+template <typename OutputType, typename InputType, int Dims>
+void Fft1DImplFfts<OutputType, InputType, Dims>::Run(
+    KernelContext &context,
+    const OutTensorCPU<OutputType, Dims> &out,
+    const InTensorCPU<InputType, Dims> &in,
+    const FftArgs &args) {
+  const auto n = in.shape[transform_axis_];
+
+  // Those should be already calculated
+  assert(transform_axis_ >= 0);
+  assert(nfft_ > 0);
+  assert(n <= nfft_);
+  bool use_real_impl = can_use_real_impl(nfft_);
+
+  auto in_buf_sz = size_in_buf(nfft_);
+  // ffts requires 32-byte aligned memory
+  float* in_buf = context.scratchpad->template Allocate<float>(AllocType::Host, in_buf_sz, 32);
+  memset(in_buf, 0, in_buf_sz*sizeof(float));
+
+  auto out_buf_sz = size_out_buf(nfft_);
+  // ffts requires 32-byte aligned memory
+  float* out_buf = context.scratchpad->template Allocate<float>(AllocType::Host, out_buf_sz, 32);
+  memset(out_buf, 0, out_buf_sz*sizeof(float));
+
+  auto in_strides = in.shape;
+  in_strides[in_strides.size()-1] = 1;
+  for (int d = in_strides.size()-2; d >= 0; d--) {
+    in_strides[d] = in_strides[d+1] * in.shape[d+1];
+  }
+  auto out_strides = out.shape;
+  out_strides[out_strides.size()-1] = 1;
+  for (int d = out_strides.size()-2; d >= 0; d--) {
+    out_strides[d] = out_strides[d+1] * out.shape[d+1];
+  }
+
+  // Step in the transform axis
+  auto out_stride = out_strides[transform_axis_];
+  auto in_stride = in_strides[transform_axis_];
+
+  std::vector<std::pair<OutputType*, const InputType*>> slices;
```

Comment:
In general case, this is a potentially large vector which you recreate at each call.

Reply:
Fixed by introducing `ForAxis`